x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizer.java

```diff
@@ -50,20 +52,31 @@ protected List<Batch<LogicalPlan>> batches() {
         rules.add(local);
         // TODO: if the local rules haven't touched the tree, the rest of the rules can be skipped
         rules.addAll(asList(operators(), cleanup()));
-        replaceRules(rules);
-        return rules;
+        return replaceRules(rules);
     }
 
+    @SuppressWarnings("unchecked")
     private List<Batch<LogicalPlan>> replaceRules(List<Batch<LogicalPlan>> listOfRules) {
-        for (Batch<LogicalPlan> batch : listOfRules) {
+        List<Batch<LogicalPlan>> newBatches = new ArrayList<>(listOfRules.size());
+        for (var batch : listOfRules) {
             var rules = batch.rules();
-            for (int i = 0; i < rules.length; i++) {
-                if (rules[i] instanceof PropagateEmptyRelation) {
-                    rules[i] = new LocalPropagateEmptyRelation();
+            List<Rule<?, LogicalPlan>> newRules = new ArrayList<>(rules.length);
+            boolean updated = false;
+            for (var r : rules) {
+                if (r instanceof PropagateEmptyRelation) {
+                    newRules.add(new LocalPropagateEmptyRelation());
+                    updated = true;
+                } else if (r instanceof ReplaceStatsFilteredAggWithEval) {
```

Comment:
Before disabling this rule locally, some BWC tests failed. The reason being that an old node was sending a fragment to a new one; the new node optimised locally pruning away the `Aggregate` node in the logical plan, which resulted in failed queries.
Having the right test query triggered this, but made me wonder if we could make this more resilient, since no other check caught this issue.
Also, likely other rules executed locally from `LocalPlanOptimizer.operators()` could be skipped locally (pruning, logical manipulations, maybe _some_ of the pushing etc.)

Reply:
Agree, but what can we do better other than BWC checks?
An alternative would be to selectively (explicitly) add rules to the local optimization batch. The upside is **awareness** (a bwc test fails -> check the newly added rules or changing an existent one must be added to LocalOptimizer or not),  the downside is explicitly reasoning about the need on the Local node as well.

Which query made this expose the issue?
I imagine it has to be a pipeline breaker. Maybe we need to get better at testing queries with pipeline breakers.

Reply:
> Which query made this expose the issue?

All the newly added queries that use an index (non-ROW). These were creating a plan on the old nodes that still required sending a fragment to data nodes (running the "new" optimising version).

> I imagine it has to be a pipeline breaker.

Yes, but specifically for Aggs, with the intermediate attributes, that's particularly "sensitive". I think reoptimising the other  pipeline breakers wouldn't alter the output.

> what can we do better other than BWC checks?

It'd probably suffice to check that the output stays unchanged.